src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1418,7 +1417,22 @@ public void Complete(Exception ex)
                 }
             }
 
-            Output.Complete();
+            // Finalize headers
+            if (!HasResponseStarted)
+            {
+                await FireOnStarting();
+            }
+
+            // Flush headers, body, trailers...
+            if (!HasResponseCompleted)
+            {
+                if (!VerifyResponseContentLength(out var lengthException))
```

Comment:
Throwing here is problematic because besides surfacing in user code, it would also surface in StopAcceptingWritesAsync which previously never threw and doesn't have error handling.